pkg/controller/instance/instance_controller.go

```diff
@@ -348,3 +384,49 @@ func parameterDifference(old, new map[string]string) map[string]string {
 
 	return diff
 }
+
+func contains(values []string, s string) bool {
+	for _, value := range values {
+		if value == s {
+			return true
+		}
+	}
+	return false
+}
+
+func remove(values []string, s string) (result []string) {
+	for _, value := range values {
+		if value == s {
+			continue
+		}
+		result = append(result, value)
+	}
+	return
+}
+
+// maybeAddFinalizer adds the cleanup finalizer to an instance if the finalizer
+// hasn't been added yet, the instance has a cleanup plan and the cleanup plan
+// did not complete yet.
+func maybeAddFinalizer(instance *kudov1beta1.Instance) {
+	if !contains(instance.ObjectMeta.Finalizers, instanceCleanupFinalizerName) {
+		if planStatus := instance.PlanStatus(v1beta1.CleanupPlanName); planStatus != nil {
+			if !planStatus.Status.IsFinished() {
```

Comment:
could you explain why is this condition necessary? maybe even add an comment. I was immediately wondering why e.g. `isRunning` is ok to add the finalizer. I actually think that we can drop this condition altogether and just add finalizer every time it's not there and it's not deleting but I might be mistaken :)

Reply:
Sure, I'll add a comment.